ipa-core/src/helpers/transport/stream/multiplexer.rs

```diff
@@ -0,0 +1,417 @@
+use std::{
+    collections::VecDeque,
+    io,
+    io::Error,
+    pin::Pin,
+    sync::atomic::{AtomicBool, Ordering},
+    task::{Context, Poll, Waker},
+};
+
+use bytes::Bytes;
+use futures::Stream;
+
+use crate::{
+    helpers::{BytesStream, LengthDelimitedStream},
+    sync::{atomic::AtomicUsize, Arc, Mutex},
+};
+
+/// Errors received from the input stream must be replicated to reach every consumer, `Arc` allows
+/// us to do that in async world.
+type MultiplexError = Arc<io::Error>;
+
+pub trait ConsumerStream: Stream<Item = Result<Bytes, MultiplexError>> {}
+impl<S: Stream<Item = Result<Bytes, MultiplexError>>> ConsumerStream for S {}
+
+/// This function takes a stream of bytes that represent a stream of RLE-encoded records and multiplexes
+/// each record into output streams based on deterministic Round-robin strategy. For example,
+/// running this function on a stream with 5 records and requesting 5 output streams, each stream will
+/// have produce only one record, with stream 1 getting record 1, stream 2 getting record 2, etc.
+///
+/// It is guaranteed that the resulting vector has the size of `shards`.
+///
+/// ## Performance
+/// This function avoids heavy contention on the `input` by imposing the strict order in which it
+/// is polled by the resulting streams. Thus, all stream handles must be polled to make progress
+/// and drain the `input`.
+///
+/// ## Errors
+/// If the input stream produces an error, it gets sent to all shards and this stream stops producing
+/// any more data.
+pub fn multiplex<S: BytesStream>(input: S, shards: usize) -> Vec<impl ConsumerStream> {
+    let poller = Arc::new(Poller {
+        consumers: (0..shards)
+            .map(|_| ConsumerState::default())
+            .map(Mutex::new)
+            .collect(),
+        state: Mutex::new(State {
+            stream: LengthDelimitedStream::new(input),
+            buf: VecDeque::default(),
+        }),
+        next: AtomicUsize::default(),
+        closed: AtomicBool::default(),
+    });
+
+    (0..shards)
+        .map(|shard| PollingEnd {
+            index: shard,
+            poller: Arc::clone(&poller),
+        })
+        .collect()
+}
+
+/// The state for each consumer stream that polls `Poller`.
+#[derive(Default)]
+struct ConsumerState {
+    waker: Option<Waker>,
+    /// Error set by the `Poller` if the outer stream failed.
+    last_error: Option<MultiplexError>,
+}
+
+struct PollingEnd<S: BytesStream> {
+    index: usize,
+    poller: Arc<Poller<S>>,
+}
+
+impl<S: BytesStream> Stream for PollingEnd<S> {
+    type Item = Result<Bytes, MultiplexError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Short circuit to avoid contention on poller when outer stream is closed.
+        if let Poll::Ready(value) = self.poller.closed_for(self.index) {
+            return Poll::Ready(value);
+        }
+
+        if self.poller.next.load(Ordering::Acquire) == self.index {
```

Comment:
I think there is a concurrency bug here (unfortunately shuffle does not seem to be able to catch atomic issues)

* set `next = 0`
* A loads `next` and its turn to poll
* B loads `next` and tries to register waker
* A is done and updates `next` to 1
* B gets its own lock and register wakers
* B exits
* A exits

Now B never gets waken

Reply:
I was surprised that Shuttle didn't catch that but then I realized that tests don't use any concurrency. I added a few other tests and shuttle immediately found two issues (one that is not mentioned above).

They are now fixed and coverage is also improved